File: boole/core/conv.py
#############################################################################
#
# conv.py
#
# description: Compute and compare normal forms.
#
#
# Authors:
# Cody Roux
#
#
##############################################################################

from expr import *
import info


def head_beta(expr):
    """Perform the following reductions:
    App(_,Abs(x,T,t),u) --> t[u/x]
    Proj(i,(t0,...,tn)) --> ti    if i <= n
    Box(_,t,T) --> t

    And do nothing otherwise.
    
    Arguments:
    - `expr`: an arbitrary expression.
    """
    if expr.is_app() and expr.fun.is_abst():
        return subst_expr([expr.arg], expr.fun.body)
    elif expr.is_fst() and expr.expr.is_pair():
        return expr.expr.fst
    elif expr.is_snd() and expr.expr.is_pair():
        return expr.expr.snd
    elif expr.is_box():
        return expr.expr
    else:
        return expr


class ParBeta(ExprVisitor):
    """Parallel beta reduction:
    reduce all beta-redexes from the bottom-up,
    without reducing redexes created by substitution
    of a bound variable.
    """
    
    def __init__(self):
        """Take as argument the substitution function,
        the opening and abstraction functions.
        """
        ExprVisitor.__init__(self)

    def visit_const(self, expr, *args, **kwargs):
        return expr

    def visit_db(self, expr, *args, **kwargs):
        return expr

    def visit_type(self, expr, *args, **kwargs):
        return expr

    def visit_kind(self, expr, *args, **kwargs):
        return expr

    def visit_bool(self, expr, *args, **kwargs):
        return expr

    def visit_bound(self, expr, *args, **kwargs):
        var, opened = open_bound_fresh(expr)
        dom = self.visit(expr.dom, *args, **kwargs)
        body = self.visit(opened, *args, **kwargs)
        body = abstract_expr([var], body)
        return Bound(expr.binder, dom, body)

    def visit_app(self, expr, *args, **kwargs):
        #We do not normalize the evidence term, as it
        #plays no role in conversion
        fun = self.visit(expr.fun, *args, **kwargs)
        arg = self.visit(expr.arg, *args, **kwargs)
        return head_beta(App(expr.conv, fun, arg))

    def visit_pair(self, expr, *args, **kwargs):
        fst = self.visit(expr.fst, *args, **kwargs)
        snd = self.visit(expr.snd, *args, **kwargs)
        type = self.visit(expr.type, *args, **kwargs)
        return Pair(fst, snd, type)

    def visit_fst(self, expr, *args, **kwargs):
        red_expr = self.visit(expr.expr, *args, **kwargs)
        return head_beta(Fst(red_expr))

    def visit_snd(self, expr, *args, **kwargs):
        red_expr = self.visit(expr.expr, *args, **kwargs)
        return head_beta(Snd(red_expr))

    def visit_ev(self, expr, *args, **kwargs):
        #do nothing, as evidence terms have
        #no computational content.
        return expr

    def visit_sub(self, expr, *args, **kwargs):
        lhs = self.visit(expr.lhs, *args, **kwargs)
        rhs = self.visit(expr.rhs, *args, **kwargs)
        return Sub(lhs, rhs)

    def visit_box(self, expr, *args, **kwargs):
        inside = self.visit(expr.expr, *args, **kwargs)
        return head_beta(Box(expr.conv, inside, expr.type))

    def visit_mvar(self, expr, *args, **kwargs):
        return expr

    def visit_tele(self, expr, *args, **kwargs):
        consts, opened_ty = open_tele_fresh(expr)
        ty_red = [self.visit(ty, *args, **kwargs) for \
                  ty in opened_ty]
        fr_vars = [c.name for c in consts]
        ty_red = [abstract_expr(fr_vars, ty) for ty in ty_red]
        return Tele(expr.vars, ty_red)

    @info.same_info
    def visit(self, expr, *args, **kwargs):
        return expr.accept(self, *args, **kwargs)
        

def par_beta(expr):
    """Perform parallel reduction of beta-expressions
    
    Arguments:
    - `expr`:
    """
    return ParBeta().visit(expr)


def beta_norm(expr):
    """Repeat beta reduction until
    the term is unchanged. May loop!
    
    Arguments:
    - `expr`:
    """
    unred = expr
    red = par_beta(expr)
    while not red.equals(unred):
        unred = red
        red = par_beta(unred)
    return red


def unfold(names, exp, context):
    """Replace a set of constants designated by names
    in expr, by looking up their definition in the context
    
    Arguments:
    - `names`: A list of strings
    - `exp`: An expression
    - `context`: A context
    """
    exprs = []
    for name in names:
        e = context.get_from_field(name, 'defs')
        exprs.append(e)
    return sub_in(exprs, names, exp)


def unfold_once(exp, context):
    """Unfolds ALL defined constants in an expression,
    NON recursive.
    
    Arguments:
    - `exp`: an expression
    - `context`: a context
    """
    all_names = [n for n in context.defs]
    return unfold(all_names, exp, context)


def unfold_all(exp, context):
    """Unfolds ALL defined constants in an expression,
    recursively.
    
    Arguments:
    - `exp`: an expression
    - `context`: a context
    """
    unred = exp
    all_names = [n for n in context.defs]
    red = unfold(all_names, exp, context)
    while not red.equals(unred):
        unred = red
        red = unfold(all_names, unred, context)
    return red

File: boole/core/conv.ml
(*****************************************************************

  conv.ml

 description: various conversion helpers, including computing
 beta-normal forms


 Authors:
 Jeremy Avigad
 Cody Roux


******************************************************************)

open Expr

exception ConstantUndefined of Expr.t

type reduction = Expr.t -> Expr.t

type conv = Expr.t -> Expr.t -> bool

let hd_beta_step t =
  match t with
    | App (Bound (Abst, _, _, t1), t2) ->
      subst t2 t1
    | Proj (Fst, Pair(_, _, t, _)) -> t
    | Proj (Snd, Pair(_, _, _, t)) -> t
    | Type _ | Const _ | DB _
    | Bound _ | App _ | Pair _ 
    | TopLevel _ | Proj _ -> t



let rec hd_beta_norm t =
  match t with
    | App (Bound (Abst, _, _, t1), t2) ->
      hd_beta_norm (subst t2 t1)
    | Proj (Fst, Pair(_, _, t, _)) -> 
      hd_beta_norm t
    | Proj (Snd, Pair(_, _, _, t)) -> 
      hd_beta_norm t
    | Type _ | Const _ | DB _
    | Bound _ | App _ | Pair _ | Proj _ | TopLevel _ -> t



let rec unfold names m t =
  match t with
    | TopLevel (a, top, ls) when List.mem a names ->
      begin try
              let def = NMap.find a m in
              let is, _ = top in
              subst_ls is ls def
        with Not_found -> raise (ConstantUndefined t)
      end
    | TopLevel _
    | Const _ | Type _ | DB _ -> t
    | Bound (b, a, ty, tm) ->
      let ty_u, tm_u = unfold names m ty, unfold names m tm in
      Bound (b, a, ty_u, tm_u)
    | App (t1, t2) -> App (unfold names m t1, unfold names m t2)
    | Pair (a, ty, t1, t2) ->
      let ty_u = unfold names m ty in
      let t1_u, t2_u = unfold names m t1, unfold names m t2 in
      Pair (a, ty_u, t1_u, t2_u)
    | Proj (p, t) -> Proj (p, unfold names m t)


let rec beta_eq t1 t2 =
  let h1 = hd_beta_norm t1 in
  let h2 = hd_beta_norm t2 in
  begin match h1, h2 with
    | Bound(b1, _, ty1, t1), Bound(b2, _, ty2, t2) ->
      (b1 = b2) && (beta_eq ty1 ty2) && (beta_eq t1 t2)
    | Pair(_, ty1, t1, u1), Pair(_, ty2, t2, u2) ->
      (beta_eq ty1 ty2) && (beta_eq t1 t2) && (beta_eq u1 u2)
    | Proj (p1, t1), Proj (p2, t2) ->
      (p1 = p2) && (beta_eq t1 t2)
    | _, _ -> Expr.equal h1 h2
  end



let rec conv t1 t2 =
  if t1 == t2 then true
  else begin
    let h1, tl1 = get_app (hd_beta_norm t1) in
    let h2, tl2 = get_app (hd_beta_norm t2) in
    let args =
      begin try
              List.for_all2 conv tl1 tl2
        with Invalid_argument _ -> false
      end
    in
    if args then
      begin match h1, h2 with
        | Bound(Pi, _, ty1, tm1), Bound(Pi, _, ty2, tm2) ->
            (beta_eq ty1 ty2) && (conv tm1 tm2)
        | Bound _, Bound _ -> beta_eq h1 h2
        | Type s1, Type s2 -> Expr.level_leq s1 s2 
        | _, _ -> Expr.equal h1 h2
      end
    else false
  end

let reduce r = r

let check_conv c = c
